### PR TITLE
Generate ACA app infrastructure

### PR DESCRIPTION
**Pull request by davidfowl**
Making progress on https://github.com/Azure/azure-dev/issues/3292
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/2965)

### CODE REVIEW

File: src/Aspire.Hosting.Azure/Provisioning/Provisioners/AzureContainerAppsInfastructure.cs [hunk 1]
```diff
@@ -0,0 +1,455 @@
...
+
+        var containerAppEnv = new AzureBicepResource("containerAppEnv", templateString:
+            """
+            param location string
+            param tags object
+
+            var resourceToken = uniqueString(resourceGroup().id)
+
+            resource logAnalyticsWorkspace 'Microsoft.OperationalInsights/workspaces@2022-10-01' = {
+              name: 'law-${resourceToken}'
```
Comment (mitchdenny) [inline src/Aspire.Hosting.Azure/Provisioning/Provisioners/AzureContainerAppsInfastructure.cs:32]: For these "global" resources I wonder if we need some concept of the app name which we can then use when generating these resource names.

File: src/Aspire.Hosting.Azure/Provisioning/Provisioners/AzureContainerAppsInfastructure.cs [hunk 2]
```diff
@@ -0,0 +1,455 @@
...
+              }
+              tags: tags
+            }
+
+            output id string = containerAppEnvironment.id
+            output logAnalyticsWorkspaceId string = logAnalyticsWorkspace.id
+            output defaultDomain string = containerAppEnvironment.properties.defaultDomain
+            """);
+
+        containerAppEnv.Annotations.Add(new ManifestPublishingCallbackAnnotation(containerAppEnv.WriteToManifest));
```
Comment (mitchdenny) [inline src/Aspire.Hosting.Azure/Provisioning/Provisioners/AzureContainerAppsInfastructure.cs:62]: Why is this necessary? Wouldn't the string that is passed in to the AddAzureBicep(...) setup the manifest writer anyway?

File: playground/AzureStorageEndToEnd/AzureStorageEndToEnd.AppHost/aspire-dashboard-containerapp.bicep [hunk 3]
```diff
@@ -0,0 +1,52 @@
+param location string
```
Comment (mitchdenny) [inline playground/AzureStorageEndToEnd/AzureStorageEndToEnd.AppHost/aspire-dashboard-containerapp.bicep]: This file shouldn't be getting emitted. You need to check whether the manifest publishing callback is set to ignore.

File: src/Aspire.Hosting.Azure/Provisioning/Provisioners/AzureContainerAppsInfastructure.cs (hunk 2 above)
Comment (davidfowl) [reply src/Aspire.Hosting.Azure/Provisioning/Provisioners/AzureContainerAppsInfastructure.cs:62]: No resource builder here.

File: playground/AzureStorageEndToEnd/AzureStorageEndToEnd.AppHost/aspire-dashboard-containerapp.bicep (hunk 3 above)
Comment (davidfowl) [reply playground/AzureStorageEndToEnd/AzureStorageEndToEnd.AppHost/aspire-dashboard-containerapp.bicep]: Yep

File: src/Aspire.Hosting.Azure/Provisioning/Provisioners/AzureContainerAppsInfastructure.cs [hunk 4]
```diff
@@ -0,0 +1,501 @@
...
+                        }
+                        """;
+                }
+
+                // TODO: Handle other protocols
+                return "";
+            }
+
+            // Create a user assigned identity for the container app
+            var containerAppIdentity = new AzureBicepResource(p.Name + "-identity", templateString:
```
Comment (davidfowl) [inline src/Aspire.Hosting.Azure/Provisioning/Provisioners/AzureContainerAppsInfastructure.cs:273]: Review, I need to make this a singleton for now.

File: src/Aspire.Hosting.Azure/Provisioning/Provisioners/AzureContainerAppsInfastructure.cs [hunk 5]
```diff
@@ -0,0 +1,669 @@
...
+    {
+        if (val.StartsWith("${") && val.EndsWith('}'))
+        {
+            return val[2..^1];
+        }
+
+        return $"'{val}'";
+    }
+
+    // Proess expressions in environment variables (or arguments)
```
Comment (DamianEdwards) [inline src/Aspire.Hosting.Azure/Provisioning/Provisioners/AzureContainerAppsInfastructure.cs:414]:
```suggestion
    // Process expressions in environment variables (or arguments)
```

File: src/Aspire.Hosting.Azure/Provisioning/Provisioners/AzureContainerAppsInfastructure.cs [hunk 6]
```diff
@@ -0,0 +1,669 @@
...
+            }
+
+            if (value is IResourceWithConnectionString csrs)
+            {
+                return ProcessValue(csrs.ConnectionStringExpression, isSecret: true);
+            }
+
+            if (value is ParameterResource param)
+            {
+                // This gets translated to an parameter
```
Comment (DamianEdwards) [inline src/Aspire.Hosting.Azure/Provisioning/Provisioners/AzureContainerAppsInfastructure.cs:468]:
```suggestion
                // This gets translated to a parameter 
```

File: src/Aspire.Hosting.Azure/Provisioning/Provisioners/AzureContainerAppsInfastructure.cs [hunk 7]
```diff
@@ -0,0 +1,669 @@
...
+
+        if (http.Any() || https.Any())
+        {
+            // ACA only supports a single http ingress rule, so we map all http/s endpoints to the same ingress
+            // https://learn.microsoft.com/en-us/azure/container-apps/ingress-overview#http
+
+            // REVIEW: Is this correct?
+            var anyHttp2 = http.Any(e => e.Transport == "http2") || https.Any(e => e.Transport == "http2");
+            var external = http.Any(e => e.IsExternal) || https.Any(e => e.IsExternal);
+            var containerPort = resource is ProjectResource ? 8080 : 80;
```
[the remark this answers is not in the document]
Comment (davidfowl) [reply src/Aspire.Hosting.Azure/Provisioning/Provisioners/AzureContainerAppsInfastructure.cs:365]: Yes this will need some thought. It can either be passed into the app or declared on the endpoint (which is probably cleaner but duplicative).

File: src/Aspire.Hosting.Azure/Provisioning/Provisioners/AzureContainerAppsInfastructure.cs [hunk 8]
```diff
@@ -0,0 +1,725 @@
...
+        if (httpIngress is not null)
+        {
+            // We're processed the http ingress, remove it from the list
+            endpointsByContainerPort.Remove(httpIngress);
+        }
+
+        // ACA can't handle > 5 additional ports so throw if that's the case here
+        if (endpointsByContainerPort.Count > 5)
+        {
+            throw new NotSupportedException("More than 5 additional ports are not supported. See https://learn.microsoft.com/en-us/azure/container-apps/ingress-overview#tcp for more details.");
```
Comment (DamianEdwards) [inline src/Aspire.Hosting.Azure/Provisioning/Provisioners/AzureContainerAppsInfastructure.cs:416]:
```suggestion
            throw new NotSupportedException("More than 5 additional ports are not supported. See https://learn.microsoft.com/azure/container-apps/ingress-overview#tcp for more details.");
```

File: src/Aspire.Hosting.Azure/Provisioning/Provisioners/AzureContainerAppsInfastructure.cs [hunk 9]
```diff
@@ -0,0 +1,725 @@
...
+        else
+        {
+            var port = endpointsByContainerPort[0].Port;
+            sb.AppendLine(CultureInfo.InvariantCulture, $"  external: false");
+            sb.AppendLine(CultureInfo.InvariantCulture, $"  targetPort: {port}");
+            sb.AppendLine("  transport: 'tcp'");
+        }
+
+        // Add additional ports
+        // https://learn.microsoft.com/en-us/azure/container-apps/ingress-how-to?pivots=azure-cli#use-additional-tcp-ports
```
Comment (DamianEdwards) [inline src/Aspire.Hosting.Azure/Provisioning/Provisioners/AzureContainerAppsInfastructure.cs:439]:
```suggestion
        // https://learn.microsoft.com/azure/container-apps/ingress-how-to?pivots=azure-cli#use-additional-tcp-ports
```

File: src/Aspire.Hosting.Azure/Provisioning/Provisioners/AzureContainerAppsInfastructure.cs [hunk 10]
```diff
@@ -0,0 +1,725 @@
...
+        sb.AppendLine("ingress: {");
+        if (httpIngress is not null)
+        {
+            sb.AppendLine(CultureInfo.InvariantCulture, $"  external: {httpIngress.External.ToString().ToLowerInvariant()}");
+            sb.AppendLine(CultureInfo.InvariantCulture, $"  targetPort: {httpIngress.Port ?? (resource is ProjectResource ? 8080 : 80)}");
+            sb.AppendLine(CultureInfo.InvariantCulture, $"  transport: '{(httpIngress.AnyH2 ? "http2" : "http")}'");
+        }
+        else
+        {
+            var port = endpointsByContainerPort[0].Port;
```
Comment (DamianEdwards) [inline src/Aspire.Hosting.Azure/Provisioning/Provisioners/AzureContainerAppsInfastructure.cs:432]: So this is taking the first non HTTP endpoint and promoting it to be the tcp ingress?

Comment (davidfowl) [reply src/Aspire.Hosting.Azure/Provisioning/Provisioners/AzureContainerAppsInfastructure.cs:432]: Yep

File: src/Aspire.Hosting.Azure/Provisioning/Provisioners/AzureProvisioner.cs [hunk 11]
```diff
@@ -55,9 +55,10 @@ private static IResource PromoteAzureResourceFromAnnotation(IResource resource)
 
     public async Task BeforeStartAsync(DistributedApplicationModel appModel, CancellationToken cancellationToken = default)
     {
-        // TODO: Make this more general purpose
         if (executionContext.IsPublishMode)
         {
+            // TODO: Let the user pick what this is, for now just hard code azure container apps
+            await new AzureContainerAppsInfastructure(executionContext).GenerateAdditionalInfrastructureAsync(appModel, cancellationToken).ConfigureAwait(false);
```
Comment (mitchdenny) [inline src/Aspire.Hosting.Azure/Provisioning/Provisioners/AzureProvisioner.cs:61]: I'm thinking that we'll want a model where we can choose per app/resource. This is fine for a x.1 release though. Just pointing to the horizon from here.

Comment (davidfowl) [reply src/Aspire.Hosting.Azure/Provisioning/Provisioners/AzureProvisioner.cs:61]: We'll need to enable that for functions, for sure, but that can't be the default mode.

**Comment by davidfowl** [comment]
Superseded by https://github.com/dotnet/aspire/pull/5470
